### PR TITLE
[nn] Define generic MultiheadAttention and SyncBatchNorm device gates from the dispatcher

### PR DESCRIPTION
**Pull request by dwoai22**
## What does this PR solves?

Two `nn` modules (activation.py and batchnorm.py) gate behaviour on a hardcoded list of device-type strings. Both lists are hand-maintained proxies for "does this device have the kernel", replace them with a dispatcher query.

Related to this RFC: #194355

### `MultiheadAttention` (`activation.py`)

`_check_arg_device` decides whether `forward` takes the fused `torch._native_multi_head_attention` fast path, by matching a hardcoded list of device-type strings. The list is a hand-maintained proxy for "does this device have the kernel", and it needs an edit every time a backend lands.

Each vendor that registers the `_native_multi_head_attention` kernel should add their key to the check list in activation.py:1059. Instead, query the disğatcher for `aten::_native_multi_head_attention`, so the gate follows the registrations rather than tracking them by hand. After this change, also XPU can use the fastpath, which is skipped even though it had a registered kernel in native_functions.yaml.

The query also fixes a second problem the list has in the other direction: PrivateUse1 is currently admitted unconditionally, so a backend that has not registered the kernel enters the fast path and hits `NotImplementedError` instead of falling back to `F.multi_head_attention_forward`.

### `SyncBatchNorm` (`batchnorm.py`, `_functions.py`)

The gate matched `["cuda", "hpu", "xpu", privateuse1]`. `hpu` was admitted despite having no in-tree kernel (not in native_functions.yaml), while the error message it raised named only GPU, XPU and privateuse1. Query `aten::batch_norm_stats` instead, so a backend that registers the kernels is accepted and one that does not gets a clear error rather than a `NotImplementedError` from inside the collective.

Separately, the graph-capture check in `_functions.py` was gated on `torch.cuda.is_available()`, which says nothing about where the input lives. On XPU, HPU and PrivateUse1 it was therefore always `False`, so the mask that forces a device-to-host sync ran unconditionally and silently broke graph capture on every non-CUDA accelerator. Key it on the input's own device module instead. There is no device-agnostic capture-state query, `torch.cuda`, `torch.xpu` and `torch_npu` each expose `is_current_stream_capturing` but `torch.accelerator` does not, so this reads it off the device module and treats absence as "not capturing", which is correct for backends without capture support.

### Dynamo

The two `torch._C` dispatcher queries are added to `constant_fold_functions` to prevent graph breaks. `torch._C._get_privateuse1_backend_name` is already in that list for the same reason.

### Behaviour changes

- XPU reaches the MHA fast path. Note the CUDA fused kernel returns `NaN` for fully-masked rows where the SDPA fallback returns `0`, a known divergence (`test_transformers.py:445` works around it with `nan_to_num()`), now inherited by newly admitted backends.
- `meta` tensors pass the MHA gate, where the old list blocked them.
- HPU is no longer accepted by SyncBatchNorm unconditionally; it now needs a registered `batch_norm_stats` kernel like every other backend. This is the point of the change but is the one thing an existing HPU user could notice.

### Test plan

Adds `test_sync_batchnorm_unsupported_device` into `test/test_nn`.

lintrunner -a
python -m pytest test/test_transformers.py -k "TestTransformers" -q
python -m pytest test/test_jit.py -k "multi_head or multihead" -q
python -m pytest test/test_modules.py -k "MultiheadAttention or BatchNorm" -q
python -m pytest test/test_nn.py -q
python -m pytest test/dynamo/test_misc.py -q

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by albanD** [comment]
Thank you very much for sending a PR.
To avoid you spending time preparing a PR for a feature we're not sure we want to add, our usual process is to discuss on the issue before doing the PR.
In particular, make sure that the issue is marked "actionable" before sending a PR!
See https://github.com/pytorch/pytorch/wiki/The-Ultimate-Guide-to-PyTorch-Contributions for details!
